### ISSUE TITLE
Support multiple license headers

### ISSUE THREAD

**Issue by rnett** [diffplug/spotless#872]
We have a situation where we need to support multiple license headers (not in the same file, but some files will have one, some the other), because some commiters have different requirements from their companies.  Spotless doesn't currently have any support for this that I could find, but it would be nice if it was able to accept multiple licenses as OK, and probably designate one as the default, to add to new files without a license header.

**Comment by nedtwigg** [comment]
Related: #650 (disable license rule for specific file)

It's already possible (under the hood) to compose a step with a filename-specific filter, e.g. this little snippet which keeps the licenseHeader step from clobbering `package-info.java` files:

https://github.com/diffplug/spotless/blob/0842bfa67ff0a6b289fc8a924a314d194e0d50d5/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/JavaExtension.java#L159


It would be pretty easy to use this mechanism to expose something like this in the [Gradle DSL](https://github.com/diffplug/spotless/blob/0842bfa67ff0a6b289fc8a924a314d194e0d50d5/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/FormatExtension.java#L402
):

```gradle
licenseHeader('foo').skipIfPathEndsWith('src/main/thirdparty/**/*.java')
licenseHeader('bar').onlyIfPathEndsWith('src/main/thirdparty/**/*.java')
```

The problem is that `LicenseHeaderStep` has [its name hardcoded](https://github.com/diffplug/spotless/blob/0842bfa67ff0a6b289fc8a924a314d194e0d50d5/lib/src/main/java/com/diffplug/spotless/generic/LicenseHeaderStep.java#L101), and you can't have multiple steps with the same name because we use that name to determine where errors come from.

https://github.com/diffplug/spotless/blob/0842bfa67ff0a6b289fc8a924a314d194e0d50d5/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/FormatExtension.java#L282-L285

Two different solutions come to mind. Happy to take a PR for either.

## Named license header

One solution is something like this:

```gradle
licenseHeader('foo').named('licenseFoo').skipIfPathEndsWith('src/main/thirdparty/**/*.java')
licenseHeader('bar').named('licenseBar').onlyIfPathEndsWith('src/main/thirdparty/**/*.java')
```

This would be pretty easy to implement. If you left off the "named()", then you'd get the "Multiple steps with name blah" error. 

## Subformats

A broader solution would be to create the concept of a subformat, which allows you to create a step which is actually a bunch of substeps, e.g.

```gradle
googleJavaFormat()
licenseHeader('foo')
subformat 'thirdparty', {
  onlyIfPathEndsWith 'src/main/java/thirdparty/**/*.java'
  licenseHeader('lgpl')
}
```

or 

```gradle
googleJavaFormat()
licenseHeader('lgpl')
subformat 'thirdparty', {
  skipIfPathEndsWith 'src/main/java/thirdparty/**/*.java'
  licenseHeader('foo')
}
```

or 

```gradle
googleJavaFormat()
subformat 'thirdparty', {
  onlyIfPathEndsWith 'src/main/java/thirdparty/**/*.java'
  licenseHeader('lgpl')
}
subformat 'ourCode', {
  onlyIfPathEndsWith 'src/main/java/com/acme/**/*.java'
  licenseHeader('foo')
}
```

The nice thing about subformats is that they could apply to any step, not just licenses. Subformat should be backed by a new `CompositeStep` which maintains the path filters and the list of steps. Implementing the DSL part would look something like this:

https://github.com/diffplug/spotless/blob/0842bfa67ff0a6b289fc8a924a314d194e0d50d5/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessExtension.java#L206

**Comment by nedtwigg** [comment]
Interesting! I guess both ideas could be extended to filter based on content rather than path:


```gradle
licenseHeader('foo').named('licenseFoo').onlyIfContentHas('/* Foo')
licenseHeader('bar').named('licenseBar').onlyIfContentHas('/* Bar')
```

```gradle
subformat 'ourCode', {
  onlyIfContentHas '/* Copyright'
  licenseHeader('foo')
}
```

Probably should also have `onlyIfContentMatches(String regex)` (see [here](https://github.com/diffplug/spotless/blob/0842bfa67ff0a6b289fc8a924a314d194e0d50d5/lib/src/main/java/com/diffplug/spotless/generic/ReplaceRegexStep.java#L34) for how we've handled regexes elsewhere).

**Comment by nedtwigg** [comment]
The key thing is single-responsibility principle. I think `LicenseHeaderStep` should *only* handle one kind of license, nothing else, and then it can be composed with other pieces to create other effects.

We've had previous requests for license-per-path, and your usecase of license-per-content is also valid. I'm open to your DSL idea, but it needs to handle the per-path and per-content cases, and they shouldn't require `LicenseHeaderStep` to change. It also shouldn't make the default case of "I want one license for all my files" any harder.

**Comment by simonbasle** [comment]
I like the subformat idea with path AND content-based filters! Such a feature could also be used to avoid overwriting a license header from a third-party file that is integrated into the codebase, untouched.

**Comment by nedtwigg** [comment]
Published in `plugin-gradle/6.0.1`.
